framework/src/interfaces/ScalarCoupleable.C

```diff
@@ -143,6 +148,54 @@ ScalarCoupleable::coupledScalarValue(const std::string & var_name, unsigned int
   return (_sc_is_implicit) ? var->sln() : var->slnOld();
 }
 
+template <ComputeStage compute_stage>
+const ADVariableValue &
+ScalarCoupleable::adCoupledScalarValueTempl(const std::string & var_name, unsigned int comp)
+{
+  checkVar(var_name);
+  if (!isCoupledScalar(var_name, comp))
+    return *getADDefaultValue<compute_stage>(var_name);
+
+  MooseVariableScalar * var = getScalarVar(var_name, comp);
+
+  if (_sc_is_implicit)
+    return var->adSln<compute_stage>();
+  else
+    mooseError("adCoupledValue for non-implicit calculations is not currently supported. Use "
+               "coupledValue instead for non-implicit");
+}
+
+template <>
+VariableValue *
+ScalarCoupleable::getADDefaultValue<ComputeStage::RESIDUAL>(const std::string & var_name)
+{
+  std::map<std::string, VariableValue *>::iterator default_value_it = _default_value.find(var_name);
+  if (default_value_it == _default_value.end())
+  {
+    VariableValue * value = new VariableValue(_sc_fe_problem.getMaxScalarOrder(),
```

Comment:
I guess we are propagating a pattern here? No `auto`, raw memory allocation etc.? This should probably all be modernized, but I suppose this is OK as a first take. It looks like you didn't forget to clean up memory 😉